### PR TITLE
[feature][update] Add UI to cancel pending registrations

### PR DESCRIPTION
**Pull request by mfraezz**
Updates #5160 

## Purpose
Allow admin contributors on pending registrations to cancel the registration through the UI, rather than having the link only be accessible through an email.

## Changes
* Additional key `'disapproval_link'` added to `'node'` dict on `_view_project` data serialization 
    * Blank if user doesn't have `ADMIN` pemissions
* UI component added -- cancel button
* Opportunistic code cleaning

UI:
screen shot 2016-04-12 at 16 58 51
screen shot 2016-04-12 at 16 59 11

## Side effects
None

## Ticket

[\[OSF-4944\]](https://openscience.atlassian.net/browse/OSF-4944)

### CODE REVIEW

File: tests/test_serializers.py [hunk 1]
```diff
@@ -158,11 +158,15 @@ def test_serialize_node_search_returns_only_visible_contributors(self):
...
+        super(TestViewProject, self).setUp()
+        self.user = UserFactory()
+        self.node = ProjectFactory(creator=self.user)
+
     # related to https://github.com/CenterForOpenScience/openscienceframework.org/issues/1109
     def test_view_project_pointer_count_excludes_folders(self):
-        user = UserFactory()
         pointer_project = ProjectFactory(is_public=True)  # project that points to another project
-        pointed_project = ProjectFactory(creator=user)  # project that other project points to
+        pointed_project = self.node  # project that other project points to
```
Comment (brianjgeiger) [inline tests/test_serializers.py]: :+1:

File: tests/test_serializers.py [hunk 2]
```diff
@@ -173,6 +177,26 @@ def test_view_project_pointer_count_excludes_folders(self):
...
+    def test_view_project_pending_registration_for_admin_user_does_contain_cancel_link(self):
+        pending_reg = RegistrationFactory(project=self.node, archive=True)
+        assert_true(pending_reg.is_pending_registration)
+        result = _view_project(pending_reg, Auth(self.user))
+
+        assert_not_equal(result['node']['disapproval_link'], '')
+        assert_in('/?token=', result['node']['disapproval_link'])
+        pending_reg.remove()
+
+    def test_view_project_pending_registration_for_admin_user_does_not_contain_cancel_link(self):
```
Comment (brianjgeiger) [inline tests/test_serializers.py]: It seems like these tests indicate contradictory behavior. My guess is that means there's a condition that causes it to sometimes contain a link and sometimes not. Be sure that the condition is in the title.

Comment (brianjgeiger) [inline tests/test_serializers.py]: Ah, it's because it should be write user not admin user, yes?

File: website/project/views/node.py [hunk 3]
```diff
@@ -660,6 +660,7 @@ def _view_project(node, auth, primary=False):
                 status.push_status_message(message, kind='info', dismissible=False, trust=True)
     data = {
         'node': {
+            'disapproval_link': node.registration_approval.stashed_urls.get(user._id, {}).get('reject', '') if (node.is_pending_registration and node.has_permission(user, ADMIN)) else '',
```
Comment (brianjgeiger) [inline website/project/views/node.py]: This is a little more than I usually like to put on a single line. How about breaking up some of the data gathering into the method above and keeping this line a little simpler? Not a big deal.

**Comment by brianjgeiger** [comment]
Looks good, seems to work, has tests. A small style issue and a renamed test and that should do it. :penguin:

File: tests/test_serializers.py (hunk 2 above)
Comment (mfraezz) [inline tests/test_serializers.py]: Correct -- write users should not receive a `'disapproval_link'`, admins should.

Comment (brianjgeiger) [inline tests/test_serializers.py]: Cool, then changing the name of the test will work.

Comment (mfraezz) [inline tests/test_serializers.py]: They're named correctly. `[...]_does_contain_cancel_link` tests to make sure that the disapproval link is there, `[...]_does_not_contain_cancel_link` tests to make sure that it's not.

Comment (brianjgeiger) [inline tests/test_serializers.py]: `test_view_project_pending_registration_for_admin_user_does_not_contain_cancel_link` should be `test_view_project_pending_registration_for_write_user_does_not_contain_cancel_link`

**Comment by mfraezz** [comment]
Also pulled and resolved a merge conflict. Back to you @brianjgeiger

**Comment by brianjgeiger** [comment]
Cool. RTM when Travis passes. :tada: :penguin: :tada:

**Comment by brianjgeiger** [comment]
I did a quick pass for Knockout Punches, and it looks like this is clean. I think it's all mako in any case.

**Pull request merged**